cpp/tests/traversal/ms_bfs_test.cpp

```diff
@@ -0,0 +1,298 @@
+#include <cugraph/graph.hpp>
+#include <cugraph/graph_functions.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/high_res_timer.hpp>
+
+#include <raft/core/handle.hpp>
+#include <raft/util/cudart_utils.hpp>
+
+#include <rmm/device_scalar.hpp>
+#include <rmm/device_vector.hpp>
+
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <iterator>
+#include <limits>
+#include <numeric>
+#include <random>
+#include <set>
+#include <vector>
+
+struct MsBfs_Usecase {
+  size_t radius;
+  size_t max_seeds;
+  bool test_weighted_{false};
+  bool check_correctness_{true};
+};
+
+template <typename input_usecase_t>
+class Tests_MsBfs : public ::testing::TestWithParam<std::tuple<MsBfs_Usecase, input_usecase_t>> {
+ public:
+  Tests_MsBfs() {}
+
+  static void SetUpTestCase() {}
+  static void TearDownTestCase() {}
+
+  virtual void SetUp() {}
+  virtual void TearDown() {}
+
+  template <typename vertex_t, typename edge_t, typename weight_t>
+  void run_current_test(std::tuple<MsBfs_Usecase const&, input_usecase_t const&> const& param)
+  {
+    constexpr bool renumber             = false;
+    auto [MsBfs_usecase, input_usecase] = param;
+    raft::handle_t handle{};
+
+    HighResTimer hr_timer{};
+
+    if (cugraph::test::g_perf) {
+      RAFT_CUDA_TRY(cudaDeviceSynchronize());  // for consistent performance measurement
+      hr_timer.start("SG Construct graph");
+    }
+
+    // NX MsBfs is not implemented for graph with self loop and multi edges therefore dropped
+    // them especially for rmat generated graphs.
+    auto [graph, edge_weights, d_renumber_map_labels] =
+      cugraph::test::construct_graph<vertex_t, edge_t, weight_t, false, false>(
+        handle, input_usecase, MsBfs_usecase.test_weighted_, renumber, true, true);
+
+    if (cugraph::test::g_perf) {
+      RAFT_CUDA_TRY(cudaDeviceSynchronize());  // for consistent performance measurement
+      hr_timer.stop();
+      hr_timer.display_and_clear(std::cout);
+    }
+
+    auto graph_view = graph.view();
+
+    auto edge_weight_view =
+      edge_weights ? std::make_optional((*edge_weights).view()) : std::nullopt;
+
+    rmm::device_uvector<vertex_t> d_labels(graph_view.number_of_vertices(), handle.get_stream());
+
+    ASSERT_TRUE(graph_view.is_symmetric())
+      << "Weakly connected components works only on undirected (symmetric) graphs.";
+
+    // Call WCC
+    cugraph::weakly_connected_components(handle, graph_view, d_labels.data());
+
+    auto d_vertices = cugraph::test::sequence<vertex_t>(
+      handle, graph_view.number_of_vertices(), size_t{1}, vertex_t{0});
+
+    rmm::device_uvector<vertex_t> d_sorted_vertices{0, handle.get_stream()};
+    rmm::device_uvector<vertex_t> d_sorted_labels{0, handle.get_stream()};
+
+    std::tie(d_sorted_labels, d_sorted_vertices) =
+      cugraph::test::sort_by_key<vertex_t, vertex_t>(handle, d_labels, d_vertices);
+
+    rmm::device_uvector<vertex_t> d_tmp_sorted_labels(d_sorted_labels.size(), handle.get_stream());
```

Comment:
Why did you create d_tmp_sorted_labels? Can we just use d_sorted_labels? We are just making a copy...